pylint/extensions/check_mccabe.py

```diff
@@ -0,0 +1,210 @@
+# -*- coding: utf-8 -*-
+"""Module to add McCabe checker class for pylint. """
+
+import ast
+
+from mccabe import McCabeChecker, PathGraph as PathGraphAstroid, PathGraphingAstVisitor
+from pylint.checkers.base import BaseChecker
+from pylint.checkers.utils import check_messages
+from pylint.interfaces import HIGH, IAstroidChecker
+
+
+class PathGraph(PathGraphAstroid):
+    def __init__(self, node):
+        super(PathGraph, self).__init__(name='', entity='', lineno=1)
+        self.root = node
+
+
+class McCabeASTVisitor(PathGraphingAstVisitor):
+    def __init__(self):
+        super(McCabeASTVisitor, self).__init__()
+        self.node = None
+        self._cache = {}
+        self.classname = ""
+        self.graphs = {}
+        self.reset()
+        self._bottom_counter = 0
+
+    def default(self, node, *args):
+        pass
+
+    def dispatch(self, node, *args):
+        self.node = node
+        klass = node.__class__
+        meth = self._cache.get(klass)
+        if meth is None:
+            className = klass.__name__
+            meth = getattr(self.visitor, 'visit' + className, self.default)
+            self._cache[klass] = meth
+        return meth(node, *args)
+
+    def visitFunctionDef(self, node):
+        if self.graph is not None:
+            # closure
+            pathnode = self._append_node(node)
+            self.tail = pathnode
+            self.dispatch_list(node.body)
+            bottom = "%s" % self._bottom_counter
+            self._bottom_counter += 1
+            self.graph.connect(self.tail, bottom)
+            self.graph.connect(node, bottom)
+            self.tail = bottom
+        else:
+            self.graph = PathGraph(node)
+            self.tail = node
+            self.dispatch_list(node.body)
+            self.graphs["%s%s" % (self.classname, node.name)] = self.graph
+            self.reset()
+
+    visitAsyncFunctionDef = visitFunctionDef
+
+    def preorder(self, tree, visitor, *args):
```

Comment:
I don't understand what's happening here, but preorder seems to be called all the time with the same visitor. Why can't this simply be self.dispatch(tree, *args)?